process/peer/validatorsProvider.go

```diff
@@ -1,69 +1,180 @@
+}
+
+// ArgValidatorsProvider contains all parameters needed for creating a validatorsProvider
+type ArgValidatorsProvider struct {
+	NodesCoordinator                  process.NodesCoordinator
+	StartEpoch                        uint32
+	EpochStartEventNotifier           process.EpochStartEventNotifier
+	CacheRefreshIntervalDurationInSec time.Duration
+	ValidatorStatistics               process.ValidatorStatisticsProcessor
+	MaxRating                         uint32
+	PubKeyConverter                   state.PubkeyConverter
 }
 
 // NewValidatorsProvider instantiates a new validatorsProvider structure responsible of keeping account of
 //  the latest information about the validators
 func NewValidatorsProvider(
-	validatorStatisticsProcessor process.ValidatorStatisticsProcessor,
-	maxRating uint32,
-	pubkeyConverter state.PubkeyConverter,
+	args ArgValidatorsProvider,
 ) (*validatorsProvider, error) {
-	if check.IfNil(validatorStatisticsProcessor) {
+	if check.IfNil(args.ValidatorStatistics) {
 		return nil, process.ErrNilValidatorStatistics
 	}
-	if check.IfNil(pubkeyConverter) {
+	if check.IfNil(args.PubKeyConverter) {
 		return nil, process.ErrNilPubkeyConverter
 	}
-	if maxRating == 0 {
+	if check.IfNil(args.NodesCoordinator) {
+		return nil, process.ErrNilNodesCoordinator
+	}
+	if check.IfNil(args.EpochStartEventNotifier) {
+		return nil, process.ErrNilEpochStartNotifier
+	}
+	if args.MaxRating == 0 {
 		return nil, process.ErrMaxRatingZero
 	}
+	if args.CacheRefreshIntervalDurationInSec <= 0 {
+		return nil, process.ErrInvalidCacheRefreshIntervalInSec
+	}
+
+	currentContext, cancelfunc := context.WithCancel(context.Background())
 
 	validatorsProvider := &validatorsProvider{
-		mutCachedMap:        sync.Mutex{},
-		cachedMap:           make(map[string]*state.ValidatorApiResponse),
-		maxRating:           maxRating,
-		validatorStatistics: validatorStatisticsProcessor,
-		pubkeyConverter:     pubkeyConverter,
+		nodesCoordinator:             args.NodesCoordinator,
+		validatorStatistics:          args.ValidatorStatistics,
+		cache:                        make(map[string]*state.ValidatorApiResponse),
+		cacheRefreshIntervalDuration: args.CacheRefreshIntervalDurationInSec,
+		refreshCache:                 make(chan uint32),
+		mutCache:                     sync.RWMutex{},
+		cancelFunc:                   cancelfunc,
+		maxRating:                    args.MaxRating,
+		pubkeyConverter:              args.PubKeyConverter,
 	}
 
+	go validatorsProvider.startRefreshProcess(currentContext, args.StartEpoch)
+	args.EpochStartEventNotifier.RegisterHandler(validatorsProvider.epochStartEventHandler())
+
 	return validatorsProvider, nil
 }
 
 // GetLatestValidators gets the latest configuration of validators from the peerAccountsTrie
 func (vp *validatorsProvider) GetLatestValidators() map[string]*state.ValidatorApiResponse {
-	vp.mutCachedMap.Lock()
-	defer vp.mutCachedMap.Unlock()
+	vp.mutCache.Lock()
+	defer vp.mutCache.Unlock()
```

Comment:
This is an exported function and you are returning a map also used internally without it's mutex, so if the caller accesses the items without locking the same mutex we might get race conditions.
Can we return a map copy instead?